### DISCUSSION TITLE
Retain types using custom argument(s)

### DISCUSSION THREAD

**Discussion by carlocorradini** [prisma/prisma#16018] [Q&A] [answered]
I don't know if this is a `TypeScript` or `Prisma` related thing, but when creating a derived argument (even a simple one), crucial type information is lost.
Take, for example, a custom type where we omit only the `include` parameter (pretty simple):
```ts
// Basic User type for example
type User = {
  id: number
  username: string
  password: string
};

// Custom argument(s)
type CustomFindUniqueArgs = Omit<Prisma.UserFindUniqueArgs, 'include'>;
```

If I try to use the custom type in a function (e.g., service class), some/all important types are lost:
```ts
function customFindUnique(args: CustomFindUniqueArgs ) {
  return prisma.status.findUnique(args);
}

function customFindUniqueObj(args: CustomFindUniqueArgs ) {
  return prisma.status.findUnique({
    where: args.where,
    select: args.select
  });
}

// Select 'username' only
const user = await customFindUnique({ where: { id: 0 }, select: { username: true } });
if (user) {
  const { id, username, password } = user; // No error! Selection is lost, only 'username' should be valid
}

// Select 'username' only | Obj
const user = await customFindUnique({ where: { id: 0 }, select: { username: true } });
if (user) {
  const { id, username, password } = user; // Error! Property 'username' does not exist on type '{}'. Types are completely lost
}
```

Is there a possible solution to this or not? Thanks 🥳

**Comment by pantharshit00** [comment, marked as answer]
Hello @carlocorradini,

The type will widen when you will define the function customFindUnique as TS will use whole UserFindUniqueArgs as it doesn’t know what will be populated in the generic. So it won't be able to give you proper autocomplete. You will need to copy almost the whole signature so that types that narrow down the type to match our needs are in place(which are `SelectSubset` along with `extends`).

So you will need to do something like this:
```ts
function customFindUnique<T extends CustomFindUniqueArgs>(
  args: Prisma.SelectSubset<T, CustomFindUniqueArgs>
): Prisma.Prisma__UserClient<Prisma.UserGetPayload<T> | null, never> {
  return prisma.user.findUnique(args);
}
```

Let me know if this works for you.

**Reply by carlocorradini** [reply]
A little bit cumbersome but it works. Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
